cmd/ferretdb/main.go

```diff
@@ -395,6 +420,20 @@ func run() {
 		stop()
 	}()
 
+	otelShutdown := setupTracer(logger)
+
+	if otelShutdown != nil {
+		defer func() {
+			shutdownCtx, shutdownCancel := ctxutil.WithDelay(ctx.Done(), 3*time.Second) //nolint:mnd // simple shutdown timeout
```

Comment:
@AlekSi I added what you had suggested, but to me, it doesn't make sense. This function is already defered, it will be executed at the moment of exiting `main`. There are no `os.Exit` starting from here, it means, exit only happens after `ctx.Done` had been done, there is no way to do it diffrenetly. So, keeping `ctx.Done()` here is unnecessary, this information won't bring any value. A simple `WithTImeout` was enough.

Reply:
That's not what I suggested, however.

![CleanShot 2024-06-27 at 08 18 37@2x](https://github.com/FerretDB/FerretDB/assets/11512/5e6ae677-eac2-42cd-9fb8-014fef2d97f7)

There is not Run method

Reply:
Could you please push what you suggested? I don't understand the idea.

Reply:
`debug.RunHandler` runs the debug handler until the passed context is canceled.
`runTelemetryReporter` runs the telemetry reporter until the passed context is canceled.
`acceptLoop` runs the loop until the passed context is canceled (with a small delay).
OTel exporter uses start and stop functions instead and has problems with context. We should change it to use the Run function/method instead. The closest analogue is `acceptLoop`.

That's just a restatement of what I wrote [there](https://github.com/FerretDB/FerretDB/pull/4380#discussion_r1652985725).

Reply:
Sorry, I think I'm too confused. Do you mean, we should write a custom otel exporter instead of using the oltp one?

Reply:
Does `debug.RunHandler` run a custom HTTP server instead of using `net/http`?

```go
// RunOtelExpoter runs OpenTelemetry exporter until the context is canceled.
func RunOtelExpoter(ctx context.Context, ...) {
```

Reply:
Well, `debug.RunHandler` has something that is run until it's stopped :) 
I don't understand what should be run in this case. Just a gorotine that waits to call `otelShutdown`? Something like this?

```
go func() {
    <-ctx.Done()
   otelShutdown(...)
}
```

Reply:
We would no longer need `otelShutdown` (and `SetupOtel`) as separate functions.

The context passed to the TracerProvider.Shutdown method should be similar to https://github.com/FerretDB/FerretDB/blob/300fbc35ef965f282e19d0ecdfca8913093eeebb/internal/clientconn/listener.go#L274-L276

Reply:
Ah, I think I got it! You want to move this logic to the `observability` package, right?

Reply:
Ether that, or it could be done the same way as `runTelemetryReporter` in the `main.go` and constructor and Run methods in their own package